src/server/qgis_mapserver.cpp

```diff
@@ -0,0 +1,382 @@
+     */
+    QString message( )
+    {
+      return mMessage;
+    }
+
+  private:
+
+    QString mMessage;
+
+};
+
+int main( int argc, char *argv[] )
+{
+  // Test if the environ variable DISPLAY is defined
+  // if it's not, the server is running in offscreen mode
+  // Qt supports using various QPA (Qt Platform Abstraction) back ends
+  // for rendering. You can specify the back end to use with the environment
+  // variable QT_QPA_PLATFORM when invoking a Qt-based application.
+  // Available platform plugins are: directfbegl, directfb, eglfs, linuxfb,
+  // minimal, minimalegl, offscreen, wayland-egl, wayland, xcb.
+  // https://www.ics.com/blog/qt-tips-and-tricks-part-1
+  // http://doc.qt.io/qt-5/qpa.html
+  const char *display = getenv( "DISPLAY" );
+  bool withDisplay = true;
+  if ( !display )
+  {
+    withDisplay = false;
+    qputenv( "QT_QPA_PLATFORM", "offscreen" );
+    QgsMessageLog::logMessage( "DISPLAY not set, running in offscreen mode, all printing capabilities will not be available.", "Server", Qgis::Info );
+  }
+  // since version 3.0 QgsServer now needs a qApp so initialize QgsApplication
+  QgsApplication app( argc, argv, withDisplay, QString(), QStringLiteral( "QGIS Development Server" ) );
+
+  QCoreApplication::setOrganizationName( QgsApplication::QGIS_ORGANIZATION_NAME );
+  QCoreApplication::setOrganizationDomain( QgsApplication::QGIS_ORGANIZATION_DOMAIN );
+  QCoreApplication::setApplicationName( "QGIS Development Server" );
+  QCoreApplication::setApplicationVersion( "1.0" );
+
+#ifdef Q_OS_WIN
+  // Initialize font database before fcgi_accept.
+  // When using FCGI with IIS, environment variables (QT_QPA_FONTDIR in this case) are lost after fcgi_accept().
+  QFontDatabase fontDB;
+#endif
+
+  // The port to listen
+  QString serverPort { getenv( "QGIS_SERVER_PORT" ) };
+  // The address to listen
+  QString ipAddress { getenv( "QGIS_SERVER_ADDRESS" ) };
+
+  if ( serverPort.isEmpty() )
+  {
+    serverPort = QStringLiteral( "8000" );
+  }
+
+  if ( ipAddress.isEmpty() )
+  {
+    ipAddress = QStringLiteral( "localhost" );
+  }
+
+  QCommandLineParser parser;
+  parser.setApplicationDescription( QObject::tr( "QGIS Development Server" ) );
+  parser.addHelpOption();
+  parser.addVersionOption();
+  parser.addPositionalArgument( QStringLiteral( "addressAndPort" ),
+                                QObject::tr( "listen to address and port (default: \"localhost:8000\")\n"
+                                    "address and port can also be specified with the environment\n"
+                                    "variables QGIS_SERVER_ADDRESS and QGIS_SERVER_PORT" ), QStringLiteral( "[address:port]" ) );
+  QCommandLineOption logLevelOption( "l", QObject::tr( "Set log level (default: 0)\n"
+                                     "0: INFO\n"
+                                     "1: WARNING\n"
+                                     "3: CRITICAL" ), "logLevel", "0" );
+  parser.addOption( logLevelOption );  //#spellok
```

Comment:
```suggestion
  parser.addOption( logLevelOption );
```

not needed anymore

Reply:
Thanks @3nids 
Sorry for not letting you do the magic first ;P